[WIP] Adding DETR

NielsRogge opened the pull request:
# What does this PR do?

I've made quite a lot of progress on implementing HuggingFace's version of [DETR](https://arxiv.org/abs/2005.12872). However, there are some remaining things to be discussed, mainly regarding `DetrFeatureExtractor`, which can be used to prepare images + annotations for the model. 

## What I currently have
There are 3 models defined: 

- `DetrModel`, which consists of a convolutional backbone + encoder-decoder Transformer, without any head on top.
- `DetrForObjectDetection`, which is `DetrModel` with 2 heads on top, namely a class labels classifier and a bounding box regressor.
- `DetrForSegmentation`, which is `DetrForObjectDetection` (yes you read that right, not `DetrModel`) with a mask head on top, for predicting segmentation masks.

Available notebooks:

- [inference notebook](https://colab.research.google.com/drive/1RWzoQHkGSfztcRcgTRcd3FJUDY4GVXVB?usp=sharing) of `DetrForObjectDetection`
- [fine-tuning notebook](https://drive.google.com/file/d/1NbG_DEPh2A87bpyQYvuutFXDvczkYAJ8/view?usp=sharing) - fine-tuning `DetrForObjectDetection` on a custom dataset (balloon dataset) - obtaining very good results!
- [inference notebook](https://colab.research.google.com/drive/1P-bz2ZBPNciT86gFQTl_qiD2LVPKqrSW?usp=sharing) of `DetrForSegmentation` (panoptic segmentation)

There's the feature extractor:

- `DetrFeatureExtractor`, which can be used to prepare images and annotations for the model. The API is is similar to `ViTFeatureExtractor` and `DeiTFeatureExtractor`: the input are image(s) + annotation(s), and the output is `pixel_values` and `pixel_mask`. 
- Currently, it only supports preparing data for object detection, not for panoptic segmentation. It is based on [this code](https://github.com/facebookresearch/detr/blob/a54b77800eb8e64e3ad0d8237789fcbf2f8350c5/datasets/coco.py#L17) in the original implementation. Given an image and annotations in COCO format, it turns the annotations into the format expected by DETR, followed by normalization + resizing of the image and corresponding annotations. 

## Questions

### 1: Supporting panoptic segmentation for DetrFeatureExtractor (done)
The problem is that if we also want to support panoptic segmentation, we rely on an external package named `panopticapi`, as it is used when preparing the annotations as can be seen [here](https://github.com/facebookresearch/detr/blob/a54b77800eb8e64e3ad0d8237789fcbf2f8350c5/datasets/coco_panoptic.py#L9). I don't know how we can add this dependency, because I assume that people don't want to install this package if they want to use DetrFeatureExtractor for object detection. How can I handle an optional dependency?

I think there are 2 options here: either 1) add a `task` argument to the feature extractor and raise an error if task = "panoptic segmentation" and panopticapi is not available, or 2) create two different feature extractors (one for object detection, and one for panoptic segmentation). The first option would look something like

```
if task == "panoptic_segmentation":
  if not is_panopticapi_available():
     raise ImportError("Panopticapi is required for the feature extractor.")
```
also, this can be at the init of the feature extractor, or at the call

### 2: DetrForPanopticSegmentation
`DetrForPanopticSegmentation` is a bit special in the sense that there are 2 ways to train this model, either 1) end-to-end, in which you train `DetrForObjectDetection` and the mask head altogether, or 2) in a 2-step process, in which you first train a `DetrForObjectDetection` model to predict bounding boxes + classes, and then in a second step you provide this model to DetrForPanopticSegmentation, freeze it and only train the mask head further for about 25 epochs. That's why there's a `box_model` argument in the `init` of `DetrForPanopticSegmentation`. 

Also, `DetrForObjectDetection` itself only uses the last feature map of the convolutional backbone, but `DetrForPanopticSegmentation` does not, it also uses layers 2, 3 and 4 of a ResNet in the mask head. There's an attribute `return_intermediate_layers` of `DetrConfig`, which should be set to `False` for `DetrForObjectDetection` and `True` for `DetrForPanopticSegmentation`. Currently, I set `config.return_intermediate_layers` to `True` no matter what at the `init` of `DetrForPanopticSegmentation`, but I don't know if hard coding this value is allowed.

### 3: timm support
My implementation supports any convolutional backbone of the [timm](https://github.com/rwightman/pytorch-image-models) package. Should I add a `is_timm_available` check for the model (instead of `is_torch_available`)? 

Supporting only object detection would make life easier, but as DETR also obtains very good results on panoptic segmentation, it would be good to support that too. 

Would love to hear your opinions @sgugger @patrickvonplaten @LysandreJik @patil-suraj

LysandreJik commented on src/transformers/models/detr/configuration_detr.py:
```diff
@@ -0,0 +1,211 @@
+        decoder_attention_heads (:obj:`int`, `optional`, defaults to 8):
+            Number of attention heads for each attention layer in the Transformer decoder.
+        decoder_ffn_dim (:obj:`int`, `optional`, defaults to 2048):
+            Dimensionality of the "intermediate" (often named feed-forward) layer in decoder.
```
```suggestion
            Dimension of the "intermediate" (often named feed-forward) layer in decoder.
```

LysandreJik commented on src/transformers/models/detr/configuration_detr.py:
```diff
@@ -0,0 +1,211 @@
+        decoder_ffn_dim (:obj:`int`, `optional`, defaults to 2048):
+            Dimensionality of the "intermediate" (often named feed-forward) layer in decoder.
+        encoder_ffn_dim (:obj:`int`, `optional`, defaults to 2048):
+            Dimensionality of the "intermediate" (often named feed-forward) layer in decoder.
```
```suggestion
            Dimension of the "intermediate" (often named feed-forward) layer in decoder.
```

LysandreJik commented on src/transformers/models/detr/detr_unused_code.py:
```diff
@@ -0,0 +1,95 @@
+## BELOW: utilities copied from
```
Is this file necessary?

LysandreJik commented on src/transformers/models/detr/feature_extraction_detr.py:
```diff
@@ -0,0 +1,656 @@
+            elif tensor_type == TensorType.PYTORCH:
+                if not is_torch_available():
+                    raise ImportError("Unable to convert output to PyTorch tensors format, PyTorch is not installed.")
+                import torch
```
It is already imported above

LysandreJik reviewed (commented):
Reviewed until feature extractor/modeling because I'm not sure you're finished. Let me know if that's the case and I'll review the rest.

Regarding your questions:

### panopticapi

How much of the panopticapi package is necessary? If it's only the `rgb2id` utility, then would it be possible to have it inside the library with the appropriate credits/copyright?

If the panopticapi package is used extensively for panoptic segmentation, then your first proposal (having  a `task` option to determine which packages should be available at runtime) sounds good to me, but happy to hear @sgugger, @patrickvonplaten and @patil-suraj's thoughts.

### The config attribute `return_intermediate_layers`

I'm not too keen on setting a config flag to a value manually within the model directly. I would rather this configuration value be used in a similar fashion as `output_hidden_states`: have the ability to override the configuration value when calling the `forward` method. This way the configuration property remains unchanged and the wrapping model (here `DetrForPanopticSegmentation`) can handle that configuration property as a standalone property.

I wonder if it should even be a configuration value if it should always have a certain value for some models? If some models require it to be set to `True` and some models require it to be set to `False`, what advantage is there to give the user the ability to change it? From the implementation it seems like an internal component that the user shouldn't concern themselves with.

### timm

It seems you've already added support for `is_timm_available()`, which is great. Cool to have it as an optional dependency!

@rwightman do you want to take a look at this implementation? This would add an optional dependency to `timm`, in return offering any `timm` model to be used as a backbone for DETR.

NielsRogge commented:
Thanks for the review, yes you've made it a lot more clear for me now:
* it's only `rgb2id` and `id2rgb` which are used from `panopticapi`, and it's only about 20 lines of code I see now. So indeed, we can just copy that code into the library (and cite the authors).
* however, I think it's still useful to have a `task` attribute at the init of `DetrFeatureExtractor`, because then we can do input type checking depending on the task, and it will also be useful regarding postprocessing the outputs of DETR (the task attribute is also done in `LukeTokenizer` for example).

rwightman commented on src/transformers/models/detr/modeling_detr.py:
```diff
@@ -0,0 +1,2239 @@
+
+        kwargs = {}
+        if name in ["resnet18", "resnet34", "resnet50", "resnet101"]:
+            kwargs["norm_layer"] = FrozenBatchNorm2d
```
FYI. There are quite a few other models besides those where `norm_layer` can be used to replace BatchNorm w/ FrozenBN, but unfortunately not all of them.

NielsRogge replied:
Yes, I wonder how we can make this list longer in order to support more models.

rwightman commented on src/transformers/models/detr/modeling_detr.py:
```diff
@@ -0,0 +1,2239 @@
+        self.return_intermediate_layers = return_intermediate_layers
+        self.num_channels = self.body.feature_info.channels(-1)
+
+        for name, parameter in self.body.named_parameters():
+            if not train_backbone or "layer2" not in name and "layer3" not in name and "layer4" not in name:
+                parameter.requires_grad_(False)
```
This is ResNet specific, at some point I do intend to provide a more generic freezing/requires_grad=False interface per feature level

sgugger replied:
This should all be done in a method of the model as freezing the backbone is a training artifact, so it should be controlled by a training parameter, not a config parameter.

sgugger commented on docs/source/index.rst:
```diff
@@ -290,6 +290,8 @@ Flax), PyTorch, and/or TensorFlow.
 +-----------------------------+----------------+----------------+-----------------+--------------------+--------------+
 |            DeiT             |       ❌       |       ❌       |       ✅        |         ❌         |      ❌      |
 +-----------------------------+----------------+----------------+-----------------+--------------------+--------------+
+|            Detr             |       ❌       |       ❌       |       ✅        |         ❌         |      ❌      |
```
Don't forget to add it in the main README too :-)

NielsRogge replied:
Done

sgugger commented on docs/source/model_doc/detr.rst:
```diff
@@ -0,0 +1,117 @@
+generalized to produce panoptic segmentation in a unified manner. We show that it significantly outperforms competitive
+baselines.*
+
+The original code can be found `here <https://github.com/facebookresearch/detr>`__.
```
```suggestion
This model was contributed by `nielsr <https://huggingface.co/nielsr>`__. The original code can be found `here <https://github.com/facebookresearch/detr>`__.
```

sgugger commented on docs/source/model_doc/detr.rst:
```diff
@@ -0,0 +1,117 @@
+Next, this is sent through the encoder, outputting :obj:`encoder_hidden_states` of the same shape (you can consider
+these as image features). Next, so-called **object queries** are sent through the decoder. This is a tensor of shape
+:obj:`(batch_size, num_queries, d_model)`, with :obj:`num_queries` typically set to 100 and is initialized with zeros.
+Each object query looks for a particular object in the image. Next, the decoder updates these object queries through
```
When reading this, I understand "one type of object" (so one head looks for all balloons, another for all horses etc.) I think each head is looking for "one object" but two heads can look for the same type (if for instance there are two balloons/two horses) so I would precise:

Each object query looks for a particular object in the image (two different heads can be looking for the same object types if there are several objects of the same type in the image).

NielsRogge replied:
Actually, re-reading the paper, I don't fully understand it: the input to the decoder is called tgt (target) and is initialized with zeros, and then these object queries are added at each attention layer. It's a bit confusing to me what the difference is between the input to the decoder and the query position embeddings.

NielsRogge replied:
Update: [this issue](https://github.com/facebookresearch/detr/issues/225#issuecomment-688799449) clarifies it.

sgugger commented on docs/source/model_doc/detr.rst:
```diff
@@ -0,0 +1,117 @@
+
+Tips:
+
+
```
```suggestion
```

sgugger commented on docs/source/model_doc/detr.rst:
```diff
@@ -0,0 +1,117 @@
+multiple self-attention and encoder-decoder attention layers to output :obj:`decoder_hidden_states` of the same shape:
+:obj:`(batch_size, num_queries, d_model)`. Next, two heads are added on top for object detection: a linear layer for
+classifying each object query into one of the objects or "no object", and a MLP to predict bounding boxes for each
+query.
```
Very nice step-by-step intro!

NielsRogge replied:
Thank you :)

sgugger commented on src/transformers/models/auto/configuration_auto.py:
```diff
@@ -346,10 +350,12 @@ def from_pretrained(cls, pretrained_model_name_or_path, **kwargs):
 
         List options
 
+
```
```suggestion
```

sgugger commented on src/transformers/models/auto/configuration_auto.py:
```diff
@@ -346,10 +350,12 @@ def from_pretrained(cls, pretrained_model_name_or_path, **kwargs):
             pretrained_model_name_or_path (:obj:`str` or :obj:`os.PathLike`):
                 Can be either:
 
+
```
```suggestion
```

sgugger commented on src/transformers/models/auto/configuration_auto.py:
```diff
@@ -385,6 +391,7 @@ def from_pretrained(cls, pretrained_model_name_or_path, **kwargs):
                 values. Behavior concerning key/value pairs whose keys are *not* configuration attributes is controlled
                 by the ``return_unused_kwargs`` keyword parameter.
 
+
```
```suggestion
```

sgugger commented on src/transformers/models/auto/modeling_auto.py:
```diff
@@ -551,6 +554,13 @@
     ]
 )
 
+MODEL_FOR_OBJECT_DETECTION_MAPPING = OrderedDict(
```
Create an Auto-class as well? Otherwise there is no point builiding the mapping.

sgugger commented on src/transformers/models/detr/configuration_detr.py:
```diff
@@ -0,0 +1,211 @@
+    Configuration objects inherit from :class:`~transformers.PretrainedConfig` and can be used to control the model
+    outputs. Read the documentation from :class:`~transformers.PretrainedConfig` for more information.
+
+
+
```
```suggestion
```

sgugger commented on src/transformers/models/detr/configuration_detr.py:
```diff
@@ -0,0 +1,211 @@
+            Whether auxiliary decoding losses (loss at each decoder layer) are to be used.
+        position_embedding_type (:obj:`str`, `optional`, defaults to :obj:`sine`):
+            Type of position embeddings to be used on top of the image features. One of 'sine' or 'learned'.
+        backbone (:obj:`bool`, `optional`, defaults to :obj:`resnet50`):
```
```suggestion
        backbone (:obj:`str`, `optional`, defaults to :obj:`"resnet50"`):
```

sgugger commented on src/transformers/models/detr/configuration_detr.py:
```diff
@@ -0,0 +1,211 @@
+            https://arxiv.org/abs/1909.11556>`__ for more details.
+        auxiliary_loss (:obj:`bool`, `optional`, defaults to :obj:`False`):
+            Whether auxiliary decoding losses (loss at each decoder layer) are to be used.
+        position_embedding_type (:obj:`str`, `optional`, defaults to :obj:`sine`):
```
```suggestion
        position_embedding_type (:obj:`str`, `optional`, defaults to :obj:`"sine"`):
```

sgugger commented on src/transformers/models/detr/configuration_detr.py:
```diff
@@ -0,0 +1,211 @@
+        auxiliary_loss (:obj:`bool`, `optional`, defaults to :obj:`False`):
+            Whether auxiliary decoding losses (loss at each decoder layer) are to be used.
+        position_embedding_type (:obj:`str`, `optional`, defaults to :obj:`sine`):
+            Type of position embeddings to be used on top of the image features. One of 'sine' or 'learned'.
```
```suggestion
            Type of position embeddings to be used on top of the image features. One of :obj:`"sine"` or :obj:`"learned"`.
```

sgugger commented on src/transformers/models/detr/configuration_detr.py:
```diff
@@ -0,0 +1,211 @@
+            Type of position embeddings to be used on top of the image features. One of 'sine' or 'learned'.
+        backbone (:obj:`bool`, `optional`, defaults to :obj:`resnet50`):
+            Name of convolutional backbone to use. Supports any convolutional backbone from the timm package.
+        train_backbone (:obj:`bool`, `optional`, defaults to :obj:`True`):
+            Whether to train (fine-tune) the backbone.
```
This is not a configuration parameter but a training parameter, this should not be saved inside the config.

NielsRogge replied:
So this means people should always set `requires_grad` for several layers to `False` each time they want to train the model?

When setting this parameter to True, it automatically does the following:
```
for name, parameter in self.body.named_parameters():
            if not train_backbone or "layer2" not in name and "layer3" not in name and "layer4" not in name:
                parameter.requires_grad_(False)
```

sgugger commented on src/transformers/models/detr/configuration_detr.py:
```diff
@@ -0,0 +1,211 @@
+            Relative weight of the generalized IoU loss in the object detection loss.
+        eos_coefficient (:obj:`float`, `optional`, defaults to 0.1):
+            Relative classification weight of the 'no-object' class in the object detection loss.
+
```
```suggestion
```

sgugger commented on src/transformers/models/detr/configuration_detr.py:
```diff
@@ -0,0 +1,211 @@
+        eos_coefficient=0.1,
+        **kwargs
+    ):
+        super().__init__(
+            is_encoder_decoder=is_encoder_decoder,
+            **kwargs,
+        )
```
```suggestion
        super().__init__(is_encoder_decoder=is_encoder_decoder, **kwargs)
```

sgugger commented on src/transformers/models/detr/feature_extraction_detr.py:
```diff
@@ -0,0 +1,803 @@
+from ...file_utils import TensorType, is_flax_available, is_tf_available, is_torch_available
+from ...image_utils import ImageFeatureExtractionMixin, is_torch_tensor
+from ...utils import logging
+from .modeling_detr import box_cxcywh_to_xyxy
```
This should be defined here with its counterpart.

sgugger commented on src/transformers/models/detr/feature_extraction_detr.py:
```diff
@@ -0,0 +1,803 @@
+
+
+# inspired by https://github.com/facebookresearch/detr/blob/master/util/box_ops.py
+def box_xyxy_to_cxcywh(x):
```
The name is obscure for anyone who has never done object detection. Something like `corners_to_center_format` would be way more understandable and the docstring can precise more what "corners" format and "center" format are.

NielsRogge replied:
Ok, renamed it.

sgugger commented on src/transformers/models/detr/feature_extraction_detr.py:
```diff
@@ -0,0 +1,803 @@
+# 2 functions below copied from https://github.com/cocodataset/panopticapi/blob/master/panopticapi/utils.py
+# Copyright (c) 2018, Alexander Kirillov
+# All rights reserved.
+def rgb2id(color):
```
```suggestion
def rgb_to_id(color):
```
We avoid the 2 usually

sgugger commented on src/transformers/models/detr/feature_extraction_detr.py:
```diff
@@ -0,0 +1,803 @@
+    return int(color[0] + 256 * color[1] + 256 * 256 * color[2])
+
+
+def id2rgb(id_map):
```
```suggestion
def id_to_rgb(id_map):
```
Same here.

sgugger commented on src/transformers/models/detr/feature_extraction_detr.py:
```diff
@@ -0,0 +1,803 @@
+def masks_to_boxes(masks):
+    """Compute the bounding boxes around the provided masks
+    The masks should be in format [N, H, W] where N is the number of masks, (H, W) are the spatial dimensions.
+    Returns a [N, 4] tensors, with the boxes in xyxy format
```
```suggestion
    Returns a [N, 4] tensors, with the boxes in corner format
```

sgugger commented on src/transformers/models/detr/feature_extraction_detr.py:
```diff
@@ -0,0 +1,803 @@
+    y_min = y.filled(fill_value=1e8)
+    y_min = y_min.reshape(y_min.shape[0], -1).min(-1)
+
+    # y_mask = (masks * np.expand_dims(y, axis=0))
+    # y_max = y_mask.reshape(y_mask.shape[0],-1).max(-1)
+    # y_min = y_mask.masked_fill(~(np.array(masks, dtype=bool)), 1e8).flatten(1).min(-1)
```
Clean up?

sgugger commented on src/transformers/models/detr/feature_extraction_detr.py:
```diff
@@ -0,0 +1,803 @@
+            Whether to resize the input to a certain :obj:`size`.
+        size (:obj:`int`, `optional`, defaults to 800):
+            Resize the input to the given size. Only has an effect if :obj:`do_resize` is set to :obj:`True`. If size
+            is a sequence like (w, h), output size will be matched to this. If size is an int, smaller edge of the
+            image will be matched to this number. i.e, if height > width, then image will be rescaled to (size * height
+            / width, size).
```
Use :obj:\` \` for all the formulas involving w, h, width, height, size.

sgugger commented on src/transformers/models/detr/feature_extraction_detr.py:
```diff
@@ -0,0 +1,803 @@
+    model_input_names = ["pixel_values", "pixel_mask"]
+
+    def __init__(
+        self, task="object_detection", do_resize=True, size=800, max_size=1333, do_normalize=True, image_mean=None, image_std=None, 
```
task is not documented.

sgugger commented on src/transformers/models/detr/feature_extraction_detr.py:
```diff
@@ -0,0 +1,803 @@
+
+    def __call__(
+        self,
+        images: Union[
+            Image.Image, np.ndarray, "torch.Tensor", List[Image.Image], List[np.ndarray], List["torch.Tensor"]  # noqa
+        ],
```
Make a typealias with the type as it's unreadable otherwise here.

sgugger commented on src/transformers/models/detr/feature_extraction_detr.py:
```diff
@@ -0,0 +1,803 @@
+
+            annotations (:obj:`Dict`, :obj:`List[Dict]`, `optional`):
+                The corresponding annotations in COCO format. The annotations for each image should have the following
+                format: {'image_id': image_id, 'annotations': target}, in other words a dicitionary with 2 keys:
```
```suggestion
                format: {'image_id': image_id, 'annotations': target}, in other words a dictionary with 2 keys:
```

sgugger commented on src/transformers/models/detr/feature_extraction_detr.py:
```diff
@@ -0,0 +1,803 @@
+
+        Returns:
+            :class:`~transformers.BatchFeature`: A :class:`~transformers.BatchFeature` with the following fields:
+
+
```
```suggestion

```

sgugger commented on src/transformers/models/detr/feature_extraction_detr.py:
```diff
@@ -0,0 +1,803 @@
+            if not isinstance(tensor_type, TensorType):
+                tensor_type = TensorType(tensor_type)
+
+            if tensor_type == TensorType.TENSORFLOW:
```
We're not supporting TensorFlow or Jax in any case, so I would clean this code here and jsut return an error if `tensor_type` is not PYTORCH.

sgugger commented on src/transformers/models/detr/modeling_detr.py:
```diff
@@ -0,0 +1,2239 @@
+
+    Args:
+        pred_masks (:obj:`torch.FloatTensor` of shape :obj:`(batch_size, ...)`):
+            ...
```
A documentation here would be nice.

sgugger commented on src/transformers/models/detr/modeling_detr.py:
```diff
@@ -0,0 +1,2239 @@
+                out.append((x, mask))
+            return out
+        
+        assert pixel_mask is not None
```
Please add an error message.

sgugger commented on src/transformers/models/detr/modeling_detr.py:
```diff
@@ -0,0 +1,2239 @@
+    def forward(self, pixel_values, pixel_mask):
+        x = pixel_values
+        mask = pixel_mask
+        assert mask is not None
```
No empty assert please.

sgugger commented on src/transformers/models/detr/modeling_detr.py:
```diff
@@ -0,0 +1,2239 @@
+        y_embed = mask.cumsum(1, dtype=torch.float32)
+        x_embed = mask.cumsum(2, dtype=torch.float32)
+        if self.normalize:
+            eps = 1e-6
+            y_embed = y_embed / (y_embed[:, -1:, :] + eps) * self.scale
+            x_embed = x_embed / (x_embed[:, :, -1:] + eps) * self.scale
```
```suggestion
            y_embed = y_embed / (y_embed[:, -1:, :] + 1e-6) * self.scale
            x_embed = x_embed / (x_embed[:, :, -1:] + 1e-6) * self.scale
```
If it's hard-coded, there is no point adding a line for it.

NielsRogge replied:
Isn't it useful to know that this is an epsilon value?

sgugger commented on src/transformers/models/detr/modeling_detr.py:
```diff
@@ -0,0 +1,2239 @@
+        j = torch.arange(h, device=x.device)
+        x_emb = self.column_embeddings(i)
+        y_emb = self.row_embeddings(j)
+        pos = (
+            torch.cat(
+                [
+                    x_emb.unsqueeze(0).repeat(h, 1, 1),
+                    y_emb.unsqueeze(1).repeat(1, w, 1),
+                ],
+                dim=-1,
+            )
+            .permute(2, 0, 1)
+            .unsqueeze(0)
+            .repeat(x.shape[0], 1, 1, 1)
+        )
```
Please use several lines to do the different operations as this is not readable.
```suggestion
        pos = torch.cat([x_emb.unsqueeze(0).repeat(h, 1, 1), y_emb.unsqueeze(1).repeat(1, w, 1)], dim=-1)
        pos = pos.permute(2, 0, 1)
        pos = pos.unsqueeze(0)
        pos = pos.repeat(x.shape[0], 1, 1, 1)
```

sgugger commented on src/transformers/models/detr/modeling_detr.py:
```diff
@@ -0,0 +1,2239 @@
+    Multi-headed attention from 'Attention Is All You Need' paper.
+
+    Here, we add position embeddings to the queries and keys (as explained in the DETR paper).
+
```
```suggestion
```

sgugger commented on src/transformers/models/detr/modeling_detr.py:
```diff
@@ -0,0 +1,2239 @@
+        is_cross_attention = key_value_states is not None
+        bsz, tgt_len, embed_dim = hidden_states.size()
+
+        # Added (Niels): add position embeddings to the hidden states before projecting to queries and keys
```
I don't think we need the Added (niels) in all those comments?

sgugger commented on src/transformers/models/detr/modeling_detr.py:
```diff
@@ -0,0 +1,2239 @@
+
+        attn_weights = torch.bmm(query_states, key_states.transpose(1, 2))
+
+        assert attn_weights.size() == (
+            bsz * self.num_heads,
+            tgt_len,
+            src_len,
+        ), f"Attention weights should be of size {(bsz * self.num_heads, tgt_len, src_len)}, but is {attn_weights.size()}"
```
This is not readable
```suggestion
        if attn_weights.size() != (bsz * self.num_heads, tgt_len, src_len):
            raise ValueError(f"Attention weights should be of size {(bsz * self.num_heads, tgt_len, src_len)}, but is {attn_weights.size()}")
```

sgugger commented on src/transformers/models/detr/modeling_detr.py:
```diff
@@ -0,0 +1,2239 @@
+        ), f"Attention weights should be of size {(bsz * self.num_heads, tgt_len, src_len)}, but is {attn_weights.size()}"
+
+        if attention_mask is not None:
+            assert attention_mask.size() == (
+                bsz,
+                1,
+                tgt_len,
+                src_len,
+            ), f"Attention mask should be of size {(bsz, 1, tgt_len, src_len)}, but is {attention_mask.size()}"
```
Same here.

sgugger commented on src/transformers/models/detr/modeling_detr.py:
```diff
@@ -0,0 +1,2239 @@
+
+        attn_output = torch.bmm(attn_probs, value_states)
+
+        assert attn_output.size() == (
+            bsz * self.num_heads,
+            tgt_len,
+            self.head_dim,
+        ), f"`attn_output` should be of size {(bsz, self.num_heads, tgt_len, self.head_dim)}, but is {attn_output.size()}"
```
Same here

NielsRogge replied:
Will update. This is all coming from the CookieCutter btw!

sgugger commented on src/transformers/models/detr/modeling_detr.py:
```diff
@@ -0,0 +1,2239 @@
+            self.head_dim,
+        ), f"`attn_output` should be of size {(bsz, self.num_heads, tgt_len, self.head_dim)}, but is {attn_output.size()}"
+
+        attn_output = (
+            attn_output.view(bsz, self.num_heads, tgt_len, self.head_dim)
+            .transpose(1, 2)
+            .reshape(bsz, tgt_len, embed_dim)
+        )
```
Please do one operation per line 
```suggestion
        attn_output = attn_output.view(bsz, self.num_heads, tgt_len, self.head_dim)
        attn_output = attn_output.transpose(1, 2)
        attn_output = attn_output.reshape(bsz, tgt_len, embed_dim)
```

sgugger commented on src/transformers/models/detr/modeling_detr.py:
```diff
@@ -0,0 +1,2239 @@
+    ):
+        """
+        Args:
+            hidden_states (:obj:`torch.FloatTensor`): input to the layer of shape `(seq_len, batch, embed_dim)`
```
```suggestion
            hidden_states (:obj:`torch.FloatTensor`): input to the layer of shape :obj:`(seq_len, batch, embed_dim)`
```

sgugger commented on src/transformers/models/detr/modeling_detr.py:
```diff
@@ -0,0 +1,2239 @@
+        Args:
+            hidden_states (:obj:`torch.FloatTensor`): input to the layer of shape `(seq_len, batch, embed_dim)`
+            attention_mask (:obj:`torch.FloatTensor`): attention mask of size
+                `(batch, 1, tgt_len, src_len)` where padding elements are indicated by very large negative values.
```
```suggestion
                :obj:`(batch, 1, tgt_len, src_len)` where padding elements are indicated by very large negative values.
```

sgugger commented on src/transformers/models/detr/modeling_detr.py:
```diff
@@ -0,0 +1,2239 @@
+        Args:
+            hidden_states (:obj:`torch.FloatTensor`): input to the layer of shape `(seq_len, batch, embed_dim)`
+            attention_mask (:obj:`torch.FloatTensor`): attention mask of size
+                `(batch, 1, tgt_len, src_len)` where padding elements are indicated by very large negative values.
```
```suggestion
                :obj:`(batch, 1, tgt_len, src_len)` where padding elements are indicated by very large negative values.
```

sgugger commented on src/transformers/models/detr/modeling_detr.py:
```diff
@@ -0,0 +1,2239 @@
+            hidden_states (:obj:`torch.FloatTensor`): input to the layer of shape `(seq_len, batch, embed_dim)`
+            attention_mask (:obj:`torch.FloatTensor`): attention mask of size
+                `(batch, 1, tgt_len, src_len)` where padding elements are indicated by very large negative values.
+            encoder_hidden_states (:obj:`torch.FloatTensor`): cross attention input to the layer of shape `(seq_len, batch, embed_dim)`
```
```suggestion
            encoder_hidden_states (:obj:`torch.FloatTensor`): cross attention input to the layer of shape :obj:`(seq_len, batch, embed_dim)`
```

sgugger commented on src/transformers/models/detr/modeling_detr.py:
```diff
@@ -0,0 +1,2239 @@
+                `(batch, 1, tgt_len, src_len)` where padding elements are indicated by very large negative values.
+            encoder_hidden_states (:obj:`torch.FloatTensor`): cross attention input to the layer of shape `(seq_len, batch, embed_dim)`
+            encoder_attention_mask (:obj:`torch.FloatTensor`): encoder attention mask of size
+                `(batch, 1, tgt_len, src_len)` where padding elements are indicated by very large negative values.
```
```suggestion
                :obj:`(batch, 1, tgt_len, src_len)` where padding elements are indicated by very large negative values.
```

sgugger commented on src/transformers/models/detr/modeling_detr.py:
```diff
@@ -0,0 +1,2239 @@
+        self.max_source_positions = config.max_position_embeddings
+        self.embed_scale = math.sqrt(embed_dim) if config.scale_embedding else 1.0
+
+        # if embed_tokens is not None:
+        #     self.embed_tokens = embed_tokens
+        # else:
+        #     self.embed_tokens = nn.Embedding(config.vocab_size, embed_dim, self.padding_idx)
+
+        # self.embed_positions = DetrLearnedPositionalEmbedding(
+        #     config.max_position_embeddings,
+        #     embed_dim,
+        #     self.padding_idx,
+        # )
```
To clean up?

sgugger commented on src/transformers/models/detr/modeling_detr.py:
```diff
@@ -0,0 +1,2239 @@
+        )
+        return_dict = return_dict if return_dict is not None else self.config.use_return_dict
+
+        # retrieve input_ids and inputs_embeds
+        # if input_ids is not None and inputs_embeds is not None:
+        #     raise ValueError("You cannot specify both input_ids and inputs_embeds at the same time")
+        # elif input_ids is not None:
+        #     input_shape = input_ids.size()
+        #     input_ids = input_ids.view(-1, input_shape[-1])
+        # elif inputs_embeds is not None:
+        #     input_shape = inputs_embeds.size()[:-1]
+        # else:
+        #     raise ValueError("You have to specify either input_ids or inputs_embeds")
+
+        # if inputs_embeds is None:
+        #     inputs_embeds = self.embed_tokens(input_ids) * self.embed_scale
+
+        # embed_pos = self.embed_positions(input_shape)
+
+        # # add position embeddings
+        # hidden_states = inputs_embeds + embed_pos
+        # hidden_states = self.layernorm_embedding(hidden_states)
+        # hidden_states = F.dropout(hidden_states, p=self.dropout, training=self.training)
```
To clean up?

sgugger commented on src/transformers/models/detr/modeling_detr.py:
```diff
@@ -0,0 +1,2239 @@
+        self.max_target_positions = config.max_position_embeddings
+        self.embed_scale = math.sqrt(config.d_model) if config.scale_embedding else 1.0
+
+        # don't think we need embed_tokens (output tokens) here, since we are just updating the query embeddings
+
+        # if embed_tokens is not None:
+        #     self.embed_tokens = embed_tokens
+        # else:
+        #     self.embed_tokens = nn.Embedding(config.vocab_size, config.d_model, self.padding_idx)
+
+        # self.embed_positions = DetrLearnedPositionalEmbedding(
+        #     config.max_position_embeddings,
+        #     config.d_model,
+        #     self.padding_idx,
+        # )
```
To clean up?

sgugger commented on src/transformers/models/detr/modeling_detr.py:
```diff
@@ -0,0 +1,2239 @@
+
+    def forward(
+        self,
+        # input_ids=None,
```
To clean up?

sgugger commented on src/transformers/models/detr/modeling_detr.py:
```diff
@@ -0,0 +1,2239 @@
+            intermediate = torch.stack(intermediate)
+
+        if not return_dict:
+            return tuple(
+                v
+                for v in [
+                    hidden_states,
+                    all_hidden_states,
+                    all_self_attns,
+                    all_cross_attentions,
+                    intermediate,
+                ]
+                if v is not None
+            )
```
```suggestion
            return tuple(
                v
                for v in [hidden_states, all_hidden_states, all_self_attns, all_cross_attentions, intermediate]
                if v is not None
            )
```
This fits like this.

sgugger commented on src/transformers/models/detr/modeling_detr.py:
```diff
@@ -0,0 +1,2239 @@
+
+
+@dataclass
+class BaseModelOutputWithCrossAttentionsAndIntermediateHiddenStates(BaseModelOutputWithCrossAttentions):
```
This is specific to Detr, so let's just name it `DetrXxxOutput`. This is way too long to be an acceptable name.

sgugger commented on src/transformers/models/detr/modeling_detr.py:
```diff
@@ -0,0 +1,2239 @@
+
+
+@dataclass
+class Seq2SeqModelOutputWithIntermediateHiddenStates(Seq2SeqModelOutput):
```
Same here.

sgugger commented on src/transformers/models/detr/modeling_detr.py:
```diff
@@ -0,0 +1,2239 @@
+        return self.decoder
+
+    @add_start_docstrings_to_model_forward(DETR_INPUTS_DOCSTRING)
+    @replace_return_docstrings(output_type=Seq2SeqModelOutput, config_class=_CONFIG_FOR_DOC)
```
Wrong output type here.

sgugger commented on src/transformers/models/detr/modeling_detr.py:
```diff
@@ -0,0 +1,2239 @@
+        # pixel_mask should be of shape (batch_size, height, width)
+        feature_map, mask, position_embeddings = self.backbone(pixel_values, pixel_mask)
+
+        assert mask is not None
```
No empty asserts please.

sgugger commented on src/transformers/models/detr/modeling_detr.py:
```diff
@@ -0,0 +1,2239 @@
+
+
+# taken from https://github.com/facebookresearch/detr/blob/master/models/segmentation.py
+class MaskHeadSmallConv(nn.Module):
```
Should be prefixed with Detr.

sgugger commented on src/transformers/models/detr/modeling_detr.py:
```diff
@@ -0,0 +1,2239 @@
+        return x
+
+
+class MHAttentionMap(nn.Module):
```
Should be prefixed with Detr

sgugger commented on src/transformers/models/detr/modeling_detr.py:
```diff
@@ -0,0 +1,2239 @@
+        Targets dicts must contain the key "boxes" containing a tensor of dim [nb_target_boxes, 4]. The target boxes
+        are expected in format (center_x, center_y, w, h), normalized by the image size.
+        """
+        assert "pred_boxes" in outputs
```
No empty asserts please!

sgugger commented on src/transformers/models/detr/modeling_detr.py:
```diff
@@ -0,0 +1,2239 @@
+
+        Targets dicts must contain the key "masks" containing a tensor of dim [nb_target_boxes, h, w].
+        """
+        assert "pred_masks" in outputs
```
No empty asserts please!

sgugger commented on src/transformers/models/detr/modeling_detr.py:
```diff
@@ -0,0 +1,2239 @@
+            "boxes": self.loss_boxes,
+            "masks": self.loss_masks,
+        }
+        assert loss in loss_map, f"do you really want to compute {loss} loss?"
```
This error message is not helpful.

sgugger commented on src/transformers/models/detr/modeling_detr.py:
```diff
@@ -0,0 +1,2239 @@
+
+
+# taken from https://github.com/facebookresearch/detr/blob/master/models/detr.py
+class MLP(nn.Module):
```
MLP is not the best name, and this should be prefixed by Detr.

NielsRogge replied:
Renamed it to `DetrMLPPredictionHead`.

sgugger commented on src/transformers/models/detr/modeling_detr.py:
```diff
@@ -0,0 +1,2239 @@
+# below: bounding box utilities taken from https://github.com/facebookresearch/detr/blob/master/util/box_ops.py
+
+
+def box_cxcywh_to_xyxy(x):
+    """
+    Converts a bounding box of format (center_x, center_y, width, height) to (x_0, y_0, x_1, y_1).
+    """
+
+    x_c, y_c, w, h = x.unbind(-1)
+    b = [(x_c - 0.5 * w), (y_c - 0.5 * h), (x_c + 0.5 * w), (y_c + 0.5 * h)]
+    return torch.stack(b, dim=-1)
+
+
+def box_xyxy_to_cxcywh(x):
+    """
+    Converts a bounding box of format (x_0, y_0, x_1, y_1) to (center_x, center_y, width, height).
+    """
+
+    x0, y0, x1, y1 = x.unbind(-1)
+    b = [(x0 + x1) / 2, (y0 + y1) / 2, (x1 - x0), (y1 - y0)]
+    return torch.stack(b, dim=-1)
```
Those two should be in feature_extraction (they are both used there) and imported in the modeling file. I also suggested names common mortals will understand ;-)

sgugger commented on src/transformers/models/detr/modeling_detr.py:
```diff
@@ -0,0 +1,2239 @@
+
+def generalized_box_iou(boxes1, boxes2):
+    """
+    Generalized IoU from https://giou.stanford.edu/ The boxes should be in [x0, y0, x1, y1] format.
```
```suggestion
    Generalized IoU from https://giou.stanford.edu/ The boxes should be in corner format.
```

sgugger commented on src/transformers/models/detr/test.py:
```diff
@@ -0,0 +1,9 @@
+from transformers import DetrConfig, DetrForPanopticSegmentation
```
Should be removed I imagine?

NielsRogge replied:
Yes, will do that at the end.

sgugger commented on tests/test_feature_extraction_detr.py:
```diff
@@ -0,0 +1,368 @@
+
+        self.assertEqual(
+            encoded_images.shape,
+            (
+                1,
+                self.feature_extract_tester.num_channels,
+                expected_height,
+                expected_width,
+            ),
```
```suggestion
            (1, self.feature_extract_tester.num_channels, expected_height, expected_width),
```
This all fits with the 119 char limit.

sgugger commented on tests/test_feature_extraction_detr.py:
```diff
@@ -0,0 +1,368 @@
+
+        self.assertEqual(
+            encoded_images.shape,
+            (
+                1,
+                self.feature_extract_tester.num_channels,
+                expected_height,
+                expected_width,
+            ),
```
As above, this all fits in one line.

sgugger commented on tests/test_modeling_detr.py:
```diff
@@ -0,0 +1,470 @@
+@require_torch
+class DetrModelTest(ModelTesterMixin, GenerationTesterMixin, unittest.TestCase):
+    all_model_classes = (
+        (
+            DetrModel,
+            DetrForObjectDetection,
+            DetrForPanopticSegmentation,
+        )
```
This all fits in one line.

sgugger commented on tests/test_modeling_detr.py:
```diff
@@ -0,0 +1,470 @@
+                self.assertEqual(len(cross_attentions), self.model_tester.num_hidden_layers)
+                self.assertListEqual(
+                    list(cross_attentions[0].shape[-3:]),
+                    [
+                        self.model_tester.num_attention_heads,
+                        decoder_seq_length,
+                        encoder_key_length,
+                    ],
```
This all fits in one line.

sgugger commented on tests/test_modeling_detr.py:
```diff
@@ -0,0 +1,470 @@
+            arg_names = [*signature.parameters.keys()]
+
+            if model.config.is_encoder_decoder:
+                expected_arg_names = [
+                    "pixel_values",
+                    "pixel_mask",
+                ]
```
This all fits in one line.

sgugger commented on utils/check_repo.py:
```diff
@@ -85,6 +88,9 @@
 # should **not** be the rule.
 IGNORE_NON_AUTO_CONFIGURED = [
     # models to ignore for model xxx mapping
+    "DetrEncoder",
+    "DetrDecoder",
+    "DetrDecoderWrapper",
```
This is unnecessary as all `Decoder`/`Encoder` are automatically ignored.

sgugger reviewed (commented):
Thanks for all your work on this!

NielsRogge commented on docs/source/model_doc/detr.rst:
```diff
@@ -0,0 +1,117 @@
+
+First, an image is sent through a pre-trained convolutional backbone (in the paper, the authors use
+ResNet-50/ResNet-101). Let's assume we also add a batch dimension. This means that the input to the backbone is a
+tensor of shape :obj:`(1, 3, height, width)`, assuming the image has 3 color channels (RGB). The CNN backbone outputs a
```
Ok, renamed to `batch_size` for clarity.

NielsRogge commented:
Addressed most comments. Once the draft is done, I will create a new branch, squash all commits and open up a new PR.

@sgugger how do I add the dummy objects for timm?
